Fixes ReactiveValidationObject NullReferenceException and enhances performance and thread-safety of ValidationText

thargy opened the pull request:
**What kind of change does this PR introduce?**
This pull request fixes #142 and implements #141.  There is no good reason for `ValidationText` to be mutable, as the current mutators (`Add` and `Clear`) are not used by the library and introduce potential thread-safety issues when validating asynchronously.

**What is the current behavior?**
#142 describes a bug based on the nullability of `ReactiveValidationObject.SelectInvalidPropertyValidations`, the fix is included in this PR as it is dependent on changes implemented for #141

#141 describes the benefits of making `ValidationText` immutable.  As well as thread-safety, we can provide the `ValidationText.Empty` and `ValidationText.None` singletons for improved performance and memory utilisation.  The current behaviour creates a lot of unnecessary lists.


**What is the new behavior?**
This PR introduces a pseudo-immutable version of `ValidationText` which exposes new `Create` factory methods.  The factory methods will automatically detect a `None` or `Empty` equivalent and return the new corresponding singleton instead of a new instance, reducing memory utilisation.

The underlying `List<string>` is replaced with a `string[]` to improve performance and memory utilisation.

Currently, the underlying array is not `readonly` as the `Add()` and `Clear()` mutators (now marked as obsolete) are implemented for backwards compatibility.  Using those methods will break thread-safety (which is not a breaking change).  If those methods are not used, the class is thread-safe.  

The existing constructors are marked as obsolete to encourage adoption of the factory methods.  

TODO comments are included for changes that can be made when the existing constructors/methods are finally removed in future.

**What might this PR break?**
The provided implementation should be backwards compatible, it marks certain methods/constructors obsolete which will produce warnings in consumers.

**Please check if the PR fulfills these requirements**
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

**Other information**:
Existing usages of the now obsolete constructors have been removed in the library and replaced with calls to the factory methods.  As such these frequently return one of the new singletons.  Trying to call `Add()` or `Clear()` on the singletons results in an `InvalidOperationException` to prevent more serious bugs.

Where consumers created their own `ValidationText` instances this is not a problem, as the singletons will not be in use in existing code.  However, attempting to modify `ValidationText` instances provided by the library itself may result in the new exception.  This is considered unlikely in practice.

worldbeater commented on src/ReactiveUI.Validation/Collections/ValidationText.cs:
```diff
@@ -13,79 +15,124 @@ namespace ReactiveUI.Validation.Collections
         /// <summary>
         /// Initializes a new instance of the <see cref="ValidationText"/> class.
         /// </summary>
+        [Obsolete("Calling the constructor is deprecated, please use ValidationText.Create() overload instead.")]
```
tip: consider marking the obsolete stuff with `[ExcludeFromCodeCoverage]`, this will prevent codecov from complaining.

thargy commented:
@worldbeater does this need tests adding for the Create factory?  Not sure where the best place to add such tests would be, also I'm a bit scuppered by #145 at the moment (though I can temporarily delete the UAP target to test).

worldbeater commented:
Yeah removing the UAP target and _not including this particular change in the diff_ of this PR might be a good way to go. Worth noting that the project builds successfully in our CI/CD environment so probably some packages might be missing or conflicting on your machine.

worldbeater reviewed (approved):
Would be happy to see the unit tests for the factories but in general LGTM! Thanks for this.

thargy commented:
Working on this now.

thargy commented:
After a further review, I think we should change ValidationText so it always contains at least one element. This is so that SelectMany on ValidationText enumerations always returns items, such as at:
ReactiveUI.Validation/src/ReactiveUI.Validation/Helpers/ReactiveValidationObject.cs

Line 93 in e7bc86d

 .SelectMany(validation => validation.Text) 
Further, as the main use of ValidationText is via ToSingleLine() the only empty item should be a singular one as found in Empty, the Create factories should therefore strip null/empty items. As mentioned, the only string.Empty item should appear in ValidationText.Empty. This makes the usage of ValidationText significantly more robust, as well as efficient.

This does break a couple of tests that assert that the empty validation text should have no items, but I don't believe it will break any real-world use cases.

worldbeater commented on src/ReactiveUI.Validation/ReactiveUI.Validation.csproj:
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <TargetFrameworks>MonoAndroid90;Xamarin.iOS10;Xamarin.Mac20;Xamarin.TVOS10;tizen40;netstandard2.0</TargetFrameworks>
-    <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">$(TargetFrameworks);net461;uap10.0.17763</TargetFrameworks>
+    <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">$(TargetFrameworks);net461</TargetFrameworks>
```
Worth reverting this change once you're done as we still want the CI/CD to generate NuGet packages for `uap10.0.17763`

thargy replied:
Damn it! Thought I'd reverted that, sorry, will do now.

thargy commented:
I believe this is now done!  After careful review, I added both an `Empty` and `None` singleton, this allowed backwards compatibility with the existing constructors.  I added tests for the `Create` factories, though they also cover much of the rest of the class too, they also indicate the desired functionality.

For example, `None+None = None`, `None+Empty=Empty`, `Empty+Empty=VT with 2 string.Empty items`, etc.

The PR should improve thread safety and memory utilisation quite noticeably as the None and Empty singletons are a very common use case in the codebase (and elsewhere).  This will result in a lot fewer object allocations in real-world scenarios.

worldbeater commented on src/ReactiveUI.Validation.Tests/ValidationTextTests.cs:
```diff
@@ -0,0 +1,229 @@
+// Licensed under the Apache License, Version 2.0 (the "License").
+// See the LICENSE file in the project root for more information.
```
Also, we are MIT-licensed btw, worth using the .NET Foundation license header:
```
// Copyright (c) 2020 .NET Foundation and Contributors. All rights reserved.
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for full license information.
```

worldbeater replied:
```suggestion
// Copyright (c) 2020 .NET Foundation and Contributors. All rights reserved.
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for full license information.
```

worldbeater commented on src/ReactiveUI.Validation/Collections/ValidationText.cs:
```diff
@@ -13,79 +16,237 @@ namespace ReactiveUI.Validation.Collections
-            _texts.Clear();
+            if (ReferenceEquals(this, Empty))
+            {
+                throw new InvalidOperationException("Adding to ValidationText.Empty is unsupported.");
```
Worth adjusting the error message to e.g. `Clearing ValidationText.Empty is unsupported` etc.

worldbeater commented on src/ReactiveUI.Validation/Collections/ValidationText.cs:
```diff
@@ -13,79 +16,237 @@ namespace ReactiveUI.Validation.Collections
+
+            if (ReferenceEquals(this, None))
+            {
+                throw new InvalidOperationException("Adding to ValidationText.None is unsupported.");
```
```suggestion
                throw new InvalidOperationException("Clearing ValidationText.None is unsupported.");
```

worldbeater commented:
Sorry for those boring suggestions! Just making sure we didn't miss anything. Huge thanks for the PR!

glennawatson commented:
Lets see if we can remove some of these TODOs and put them into a github issue.

glennawatson commented on src/ReactiveUI.Validation/Collections/ValidationText.cs:
```diff
@@ -13,79 +16,237 @@ namespace ReactiveUI.Validation.Collections
+        /// </summary>
+        public static readonly ValidationText Empty = new ValidationText(new[] { string.Empty });
+
+        // TODO When Add() & Clear() are obsolesced this should be made read-only.
```
```suggestion
```

thargy commented:
Thanks @worldbeater & @glennawatson for taking it over the line, I'd gone to bed!

worldbeater commented:
Should be available on NuGet as `1.8.6` now.
